aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
 #if __METAL_VERSION__ >= 310
 INSTANTIATE_INT8MM(bfloat);
 #endif
+
+// ------------------------------ For M >= 4 ------------------------------------
+/**
+ * The following code is heavily inspired by llama.cpp (https://github.com/ggerganov/llama.cpp).
+ * The original code is under MIT License: https://github.com/ggerganov/llama.cpp/blob/master/LICENSE
+ *
+ * Matrix Multiplication Algorithm:
+ * 1. Load A and B blocks (32x32 and 64x32 respectively) into shared memory.
+ * 2. In 4 simdgroups, calculate the outer product of the loaded blocks. Each simdgroup produces a 2x4 8x8 result.
```

Comment:
wikipedia is unclear about the meaning of "outer product" as it applies to matrices. are you referring to what they're calling Kronecker product?